compiler_opt/rl/corpus.py

```diff
@@ -30,9 +33,100 @@ class ModuleSpec:
   """
   name: str
   exec_cmd: Tuple[str, ...] = ()
-
-
-def build_modulespecs_from_datapath(
+  size: int = 0
+
+
+class SamplerBucketRoundRobin:
+  """Calls return a list of module_specs sampled randomly from n buckets, in
+  randomized-round-robin order. The buckets are sequential sections of
+  module_specs of roughly equal lengths."""
+
+  def __init__(self):
+    self._ranges = {}
+
+  def __call__(self,
+               module_specs: List[ModuleSpec],
+               k: int,
+               n: int = 20) -> List[ModuleSpec]:
+    """
+    Args:
+      module_specs: list of module_specs to sample from
+      k: number of modules to sample
+      n: number of buckets to use
+    """
+    # Credits to yundi@ for the highly optimized algo
+    # Essentially, split module_specs into k buckets, then define the order of
+    # visiting the k buckets such that it approximates the behaviour of having
+    # n buckets.
+    specs_len = len(module_specs)
+    if (specs_len, k, n) not in self._ranges:
+      quotient, remainder = divmod(k, n)
+      # rev_map maps from bucket # (implicitly via index) to order of visiting.
+      # lower values should be visited first, and earlier indices before later.
+      rev_map = list(range(quotient)) * n + list(range(remainder))
```

Comment:
rev_map = [x % quotient for x in range(k)] for a more even break --- see k = 119, n = 20 for example

Reply:
added a check for when k < n too